### PR TITLE
online recovery: fix online recovery timeout mechanism

### LINKED ISSUE #6107: Online recovery timeout doesn't work
### What did you do?

Run online recovery with the TiKV without the support of online recovery

### What did you expect to see?

keeps in collecting reports stage 

### What did you see instead?

timeout

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by Connor1996**
### What problem does this PR solve?

Issue Number: Close #6107 

### What is changed and how does it work?

```commit-message
fix online recovery timeout mechanism
```

### Check List

Tests

- Unit test

### Release note

```release-note
Fix the issue that online recovery timeout may not work as expected
```

### CODE REVIEW

**Comment by Connor1996** [comment]
PTAL @v01dstar

**Comment by v01dstar** [comment]
Can you please explain a little bit more about what the bug is? I can tell that with this change, the whole process will exit faster when timeout happens (the old / existing also exit after timeout, i believe? ). Besides, I think, with this change, we may leave some regions in exit force leader state when timeout?

**Comment by Connor1996** [comment]
> Can you please explain a little bit more about what the bug is? I can tell that with this change, the whole process will exit faster when timeout happens (the old / existing also exit after timeout, i believe? ). Besides, I think, with this change, we may leave some regions in exit force leader state when timeout?

Suppose that, one TiKV always returns store heartbeat but without store report. Then in the existing impl, it would never trigger timeout and keep in the collecting stage forever.

**Comment by v01dstar** [comment]
> > Can you please explain a little bit more about what the bug is? I can tell that with this change, the whole process will exit faster when timeout happens (the old / existing also exit after timeout, i believe? ). Besides, I think, with this change, we may leave some regions in exit force leader state when timeout?
> 
> Suppose that, one TiKV always returns store heartbeat but without store report for somewhat reason. Then in the existing impl, it would never trigger timeout and keep in the collecting stage forever.

I think the existing code still exit? Just with a longer wait time, `timeout + 2* store_heartbeat_interval`. Am I miss something here?

**Comment by Connor1996** [comment]
> > > Can you please explain a little bit more about what the bug is? I can tell that with this change, the whole process will exit faster when timeout happens (the old / existing also exit after timeout, i believe? ). Besides, I think, with this change, we may leave some regions in exit force leader state when timeout?
> > 
> > 
> > Suppose that, one TiKV always returns store heartbeat but without store report for somewhat reason. Then in the existing impl, it would never trigger timeout and keep in the collecting stage forever.
> 
> I think the existing code still exit? Just with a longer wait time, `timeout + 2* store_heartbeat_interval`. Am I miss something here?

Please check `checkTimeout` in existing code, it never change the state.

File: server/cluster/unsafe_recovery_controller.go [hunk 1]
```diff
@@ -274,128 +279,139 @@ func (u *unsafeRecoveryController) HandleErr(err error) bool {
...
+		// automatically after being for a long time.
 		u.changeStage(failed)
 		return true
 	}
+	// When encountering an error for the first time, we will try to exit force
+	// leader before turning into failed stage to avoid the leaking force leaders
+	// blocks reads and writes.
+	u.storePlanExpires = make(map[uint64]time.Time)
+	u.storeRecoveryPlans = make(map[uint64]*pdpb.RecoveryPlan)
+	u.timeout = time.Now().Add(storeRequestInterval)
```
Comment (v01dstar) [inline server/cluster/unsafe_recovery_controller.go:293]: Isn't one heartbeat interval too aggressive? Maybe *2 to make it more stable?

**Review by v01dstar — APPROVED** [review]

**Review by rleungx — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**